[Wee Xin Yang, Markus] iP

yellow-294 opened the pull request:
### LUNA

LUNA frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/yellow-294/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing ToDo tasks
- [x] Managing Event tasks
- [x] Managing Deadline tasks 
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

`public static void main(String[] args) throws IOException, ParseException, DukeException {
        new Duke().run();
    }`

Commit:
Add Gradle support

yellow-294 committed:
Level-1

yellow-294 committed:
Level 1. Greet, Echo, Exit

yellow-294 committed:
Completed Level-2. Add, List

yellow-294 committed:
Merge branch 'master' of https://github.com/yellow-294/ip

yellow-294 committed:
Completed Level-3 Mark As Done

yellow-294 committed:
Completed Level4.ToDos, Events, Deadlines

yellow-294 committed:
Updated A-TextUiTesting

yellow-294 committed:
Completed Level 5. Handle Errors

yellow-294 committed:
Completed Level 6. Delete

Also made changes to previous code, mainly shortening and making it more efficient

yellow-294 committed:
Completed Level-7 on separate branch

yellow-294 committed:
Completed Level-8 on separate branch

yellow-294 committed:
Merge branch 'branch-Level-8'

yellow-294 committed:
Completed A-MoreOOP

I have previously made a class with similar purposed to to "Storage", however I have named it as SavedTaskHandler.

yellow-294 committed:
Completed A-Packages

yellow-294 committed:
Ammeded problems with packages

yellow-294 committed:
Merge branch 'add-gradle-support'

yellow-294 committed:
Completed A-Gradle

yellow-294 committed:
Completed A-JUnit

yellow-294 committed:
Completed A-JavaDoc

yellow-294 committed:
Fixed unchecked warnings

yellow-294 committed:
Commit to branch-A-JavaDoc

yellow-294 committed:
Commit to branch-A-CodingStandard

yellow-294 committed:
Add branch-Level-9

yellow-294 committed:
Merge branch 'branch-A-CodingStandard'

yellow-294 committed:
Merge branch 'branch-Level-9'

wweqg commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+            this.localDate = LocalDate.parse(dateStr);
+        } catch (DateTimeParseException e) {
+            System.out.println("Please input dates in the following format: yyyy-mm-dd (eg. 2019-10-15)");
+
```
Would it be better to let Ui object to handle the display message here ? 🤔

wweqg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,44 @@
+
+
+    public static void main(String[] args) throws IOException, ParseException, DukeException {
+        new Duke().run();
```
Hi I notice the exceptions here are being thrown instead of being handle. Would these exceptions terminate the main program under certain circumstances ? 🤔

wweqg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+
+        } else if (input.substring(0, Math.min(input.length(), 4)).equals("mark")
+                || input.substring(0, Math.min(input.length(), 6)).equals("unmark")
+                || input.substring(0, Math.min(input.length(), 6)).equals("delete")) {
```
Would it be better if these numbers were to define as constants ?  For example:
```
final int START_OF_COMMAND = 0;
final int END_OF_FOUR_WORDS_COMMAND = 4;
```
Then this line `input.substring(0, Math.min(input.length(), 4)).equals("mark")` can be translated to
`input.substring(START_OF_COMMAND , Math.min(input.length(), END_OF_FOUR_WORDS_COMMAND)).equals("mark")`

wweqg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+                || input.substring(0, Math.min(input.length(), 6)).equals("unmark")
+                || input.substring(0, Math.min(input.length(), 6)).equals("delete")) {
+            String str = "";
+            boolean error = false;
```
I guess the Boolean variable name here could be `isError` ? 🤔

wweqg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+
+            try {
+
+                if (splitStr.length > 2 || splitStr.length == 1 || !isNumeric(splitStr[1])) {
```
I guess it would be better to define these numbers 👍

wweqg commented on src/main/java/duke/SavedTaskHandler.java:
```diff
@@ -0,0 +1,117 @@
+ */
+public class SavedTaskHandler {
+
+    private final String filePath = "./data/saved.txt";
```
Would this be better ? 🤔 
```
private static final String CURRENT_DIRECTORY = "user.dir";
String path = Paths.get(System.getProperty(CURRENT_DIRECTORY), "data", "saved.txt").toString();
```
I guess this way would handle potential path related run time exceptions in some operating systems.

wweqg commented on src/main/java/duke/SavedTaskHandler.java:
```diff
@@ -0,0 +1,117 @@
+                            taskList.add(events);
+                        }
+                    }
+
```
I guess it would be better to avoid deep nesting (> 3 levels of indentation) 👍

wweqg commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    public Task set(int index, Task task) {
```
Missing javadoc here 🤚

wweqg commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,74 @@
+        return this.arrayList.size();
+    }
+
+    public void forEach(Consumer<? super Task> consumer) {
```
I find this method interesting. Would you like to share the possible usage for it ? 😊

Santosh3007 replied:
You might want to add Javadocs here as well for understanding :)

wweqg commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,98 @@
+    private final String TODO_MESSAGE = "LUNA will write this one down on her finest crater...:\n";
+    private final String EVENT_MESSAGE = "What event could possibly be more important than the Moon Festival?\n";
+    private final String DEADLINE_MESSAGE = "Deadlines... LUNA isn't too good with those...\n";
+    private final String FIND_MESSAGE = "These are the Tasks you have been finding:";
```
Pretty neat and clear organization of messages here ! 👍

wweqg reviewed (commented):
Overall your program is pretty awesome. The structure and program flow are clear. I like how regex is being used in your code, it is quite efficient.

Just a side note, I guess it would be better if there were different packages to organize the classes, especially if you are going to add on more classes.

yellow-294 committed:
Add GUI

linuschancs commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ * The Deadlines class, which is a subclass of Task,
+ * encapsulates Deadlines objects
```
Missing punctuation at the end of the header comment.

linuschancs commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,13 @@
+
+public class Launcher {
+
+    /**
+     * A launcher class to workaround classpath issues.
+     */
```
These class header comments should go above the class name!

linuschancs commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+                    } else if (type.equals("find")) {
+                        throw new DukeException("Find... What exactly? ");
+                    } else {
+                        throw new DukeException("LUNA has consulted the Moon Goddess and even she "
```
Really love the personality you have given to your program!

linuschancs commented on src/main/java/duke/SavedTaskHandler.java:
```diff
@@ -0,0 +1,117 @@
+    private final String filePath = "./data/saved.txt";
+    private final File file;
+    private final Path path;
+    private final String EMPTY_TASKS = "Go ahead, your first task, tell LUNA and she will write it down...";
```
Good naming convention followed for constants!

linuschancs commented on src/main/java/duke/SavedTaskHandler.java:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * Returns the TaskList object encapsulated in this SavedTaskHandler object
+     * @return TaskList obejct
+     *
```
I noticed that for some of your methods below there is an additional empty line here that should be removed!

linuschancs reviewed (commented):
Overall really like the personality that you have given to your program and the structure of your code! Very meticulous JavaDocs has also been written for your classes and methods. Coding convention was followed for the most part and code was easy to understand as a reviewer! Really good job!

Santosh3007 commented on src/main/java/duke/Events.java:
```diff
@@ -0,0 +1,28 @@
+ * @version 0.1
+ * @since   2022-8-24
+ */
+public class Events extends Task {
```
Would it be better to name this class `Event` instead of `Events` since it represents a single event and not a group of events?

Santosh3007 replied:
Would it maybe be better to name this class `Event` instead of `Events` since it represents a single event and not a group of events?

Santosh3007 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,32 @@
+public class Main extends Application {
+    private Duke duke = new Duke();
+
+    public Main() throws IOException, ParseException {
```
Is there a reason you are throwing IOException or ParseException from this constructor since there is nothing inside the constructor? Also if your constructor is not doing anything, I think you can just remove the entire method declaration.

Santosh3007 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,60 @@
+
+    /**
+     * Creates two dialog boxes, one echoing user input and the other containing Duke's reply and then appends them to
+     * the dialog container. Clears the user input after processing.
```
You might want to include the exception your method will be throwing and why it might throw it in the javadocs.
```
/**
  * Creates two dialog boxes, one echoing user input and the other containing Duke's reply and then appends them to
  * the dialog container. Clears the user input after processing.
  * @throws ParseException if user input is invalid.
  * /

```

Santosh3007 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+
+    /**
+     * Returns a boolean value representing whether the input is Numeric
+     * @param string The string that will be parsed to check if it is Numeric
```
I think the formatting for javadocs is a bit off here. Usually there is a blank line after your method description and there is no blank lines between the block tags (Eg. @param, @return, @throws). Also, end your descriptions with a full stop.
```java
   /**
     * Returns a boolean value representing whether the input is Numeric.
     * 
     * @param string The string that will be parsed to check if it is Numeric
     * @return A boolean showing if the input string is Numeric
     */
```

Santosh3007 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+     * Parses the user's input and assign different actions according to the input
+     * @param input The string that will be parsed
+     *
+     */
```
Same thing here for Javadoc formatting and you might want to include `@throws` block tag and remove the blank line at the end of the javadocs.

Santosh3007 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,186 @@
+
+            if (!isError) {
+                if (type.equals("todo")) {
+
```
You might want to be consistent with your blank lines in your if blocks since at some places you use blank line after the braces and other places you do not. Usually, there is no blank lines directly before and after Egyptian braces.

Santosh3007 commented on src/main/java/duke/ToDos.java:
```diff
@@ -0,0 +1,26 @@
+ * @version 0.1
+ * @since   2022-8-24
+ */
+public class ToDos extends Task {
```
Same thing here. This class can be names `Todo` since it deals with a single Todo and also since it is inheriting from `Task` which is singular as well.

Santosh3007 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,93 @@
+     * Prints a message indicating a Task has been unmarked
+     */
+    public String unmarkMessage(String taskName) {
+
```
You might want to take note of the inconsistent blank lines here as well :)

Santosh3007 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Prints a message indicating a ToDos Object has been added to the TaskList
+     */
+    public String todoMessage(ToDos todo, int len) {
```
I think you missed out the `@param` block tag for all the javadocs in this file and I don't really understand what the parameter `len` is used for in this method.

Santosh3007 commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+ * @version 0.1
+ * @since   2022-8-24
+ */
+public class Deadlines extends Task {
```
Same thing for the class name here,

Santosh3007 reviewed (commented):
LGTM. Just a few minor coding standard changes needed Nice work with the user interaction also!

yellow-294 committed:
Update build.gradle

yellow-294 committed:
Complete A-Assertions

yellow-294 committed:
Complete A-CodeQuality

Improve Code Quality

There were several methods that did not use Egyptian Braces.

Changing them to use Egyptian Braces improves code quality.

yellow-294 committed:
Merge pull request #1 from yellow-294/branch-A-Assertions

Complete A-Assertions

yellow-294 committed:
Merge pull request #2 from yellow-294/branch-A-CodeQuality

Complete A-CodeQuality

yellow-294 committed:
Complete C-Priority

Add Priority marking to Tasks

This allows the user to mark each task's priority as high, medium, or low

This priority marking is done after the task has been added

yellow-294 committed:
Shorten method

Cleaned up code and made the SavedTaskHandler constructor shorter

yellow-294 committed:
Improve Code Quality

Created command Package to contain various Command classes to encapsulate various commands.

Added more Error Handling.

Added more personality.

yellow-294 committed:
Improve GUI

Implemented a better GUI with nicer colours and padding

yellow-294 committed:
Update JavaDocs and Ui.png

Added JavaDocs for public methods and constructors. Also added Ui.png into docs

yellow-294 committed:
Update README.md and implement app exit

Updated README.md to be the user guide.

Also made changes such that the 'bye' command will exit the app

yellow-294 committed:
Fix Style

yellow-294 committed:
Correct README.md

yellow-294 committed:
Fix Bugs

yellow-294 committed:
Update README.md and ammend Events

Events now takes in dates in the format YYYY-MM-DD

yellow-294 committed:
Amend Priority and SavedTaskHandler

Handled some exceptions

yellow-294 committed:
Update README.md
